crypto/sha/asm/keccak1600-mmx.pl

```diff
@@ -256,13 +253,16 @@ ()
 	&psllq	(@T[1],$rhotates[4][($x+4)%5]);
 	&por	(@C[4],@T[1]);		# C[4] = ROL64(A[4][2] ^ D[2], rhotates[4][2]);
 }
-	&Rho	(3);	&Chi	(1);
-	&Rho	(1);	&Chi	(2);
-	&Rho	(4);	&Chi	(3);
+	&Rho	(3);	&Chi	(1, 1);
+	&Rho	(1);	&Chi	(2, 4);
+	&Rho	(4);	&Chi	(3, 2);
 	&Rho	(2);	###&Chi	(4);
 
 	&movq	(@T[0],@C[0]);		######### last Chi(4) is special
+	 &xor	("edi","esi");		# &xchg	("esi","edi");
 	&movq	(&QWP(@D[1],"esp"),@C[1]);
+	 &xor	("esi","edi");
+	 &xor	("edi","esi");
```

Comment:
Cool!
Hope gcc does not generate these xchg insns when they are so bad on this cpu type.
(Just in case, I happen to be also a gcc maintainer)